arc_cache.go

```diff
@@ -151,8 +168,16 @@ func (b *arccache) PutMany(bs []blocks.Block) error {
 		// the block isn't in storage
 		if has, _, ok := b.queryCache(block.Cid()); !ok || (ok && !has) {
 			good = append(good, block)
+			b.lks[block.Cid().Bytes()[len(block.Cid().Bytes())-1]].Lock()
```

Comment:
This can deadlock if two concurrent `PutMany` calls try to put blocks in different orders. It might also try to take the same lock twice.

The simplest solution is likely to:

1. Allocate a `mxs := [256]*sync.Mutex`.
2. For each block, assign `mxs[myIndex] = &myMutex`.
3. Lock/unlock all non-nil locks.

There may be faster schemes with bit-twiddling, but this should be fine.